src/types/mod.rs

```diff
@@ -774,3 +902,178 @@ impl<T: Object> ValueInternal for JsFunction<T> {
+/// ```
+/// # use neon::prelude::*;
+/// # fn foo(mut cx: FunctionContext) -> JsResult<JsObject> {
+/// # let global = cx.global();
+/// # let url = global.get(&mut cx, "URL")?;
+/// # let url: Handle<JsFunction> = url.downcast_or_throw(&mut cx)?;
+/// let obj = url
+///     .arg(cx.string("https://neon-bindings.com"))
+///     .new(&mut cx)?;
+/// # Ok(obj)
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct Call<'a, T: Object = JsObject> {
+    callee: Handle<'a, JsFunction<T>>,
+    args: ArgsVec<'a>,
+}
+
+impl<'a, T: Object> FunctionCall<'a, T> {
+    /// Set the value of `this` for the function call.
+    pub fn this<V: Value>(&mut self, this: Handle<'a, V>) -> &mut Self {
+        self.this = this.upcast();
+        self
+    }
+
+    /// Add an argument to the arguments list.
+    pub fn arg<V: Value>(&mut self, arg: Handle<'a, V>) -> &mut Self {
+        self.args.push(arg.upcast());
+        self
+    }
+
+    /// Add multiple arguments to the arguments list.
+    pub fn args<A: Arguments<'a>>(&mut self, args: A) -> &mut Self {
+        args.append(&mut self.args);
+        self
+    }
+
+    /// Make the function call. If the function returns without throwing, the result value
+    /// is downcast to the type `V`, throwing a `TypeError` if the downcast fails.
+    pub fn call<'b, C: Context<'b>, V: Value>(&self, cx: &mut C) -> JsResult<'b, V> {
+        let v: Handle<JsValue> = self.callee.do_call(cx, self.this, &self.args)?;
+        v.downcast_or_throw(cx)
+    }
+
+    /// Make the function call for side effect, discarding the result value. This method is
+    /// preferable to [`call()`](crate::types::FunctionCall::call) when the result value is
+    /// not needed, since it does not require specifying a result type.
+    pub fn exec<'b, C: Context<'b>>(&self, cx: &mut C) -> NeonResult<()> {
+        let _: Handle<JsValue> = self.callee.do_call(cx, self.this, &self.args)?;
+        Ok(())
+    }
+}
+
+impl<'a, T: Object> Call<'a, T> {
+    /// Set the value of `this` for the function call. Once a call has a `this` binding
+    /// specified, it is required to be a [`FunctionCall`](crate::types::FunctionCall).
+    pub fn this<V: Value>(self, this: Handle<'a, V>) -> FunctionCall<'a, T> {
+        FunctionCall {
+            callee: self.callee,
+            this: this.upcast(),
+            args: self.args,
+        }
+    }
+
+    /// Add an argument to the arguments list.
+    pub fn arg<V: Value>(&mut self, arg: Handle<'a, V>) -> &mut Self {
+        self.args.push(arg.upcast());
+        self
+    }
+
+    /// Add multiple arguments to the arguments list.
+    pub fn args<A: Arguments<'a>>(&mut self, args: A) -> &mut Self {
+        args.append(&mut self.args);
+        self
+    }
+
+    /// Call the function as a constructor (like a JavaScript `new` expression).
+    /// If the function returns without throwing, returns the resulting object.
+    pub fn new<'b, C: Context<'b>>(&self, cx: &mut C) -> JsResult<'b, T> {
```

Comment:
Should we avoid the `new` keyword in case it's added as a keyword in future Rust versions?

Reply:
I don't think it'll be made into a Rust keyword, but I think we should use `.construct` for consistency.